### ISSUE TITLE
Add (optional) count of files/tags for listings

### ISSUE THREAD

**Issue by mserajnik** [mserajnik/hydrusrv#15]
Requests for file/tag listings should (optionally) include a count of the total number of items (across all pages).

SQLite doesn't have any fancy functions to count the total set when using `LIMIT` (e.g., MySQL has `SQL_CALC_FOUND_ROWS`, although it's debatable if that is actually faster than running two queries). So we need to run two queries:

+ one for the current page (with `LIMIT`)
+ one for the total count (without `LIMIT`)

This will increase query (and therefore response) times quite a bit. That's why it should be optional (and off by default); let the users decide if they want to trade performance for convenience.

~~It might be possible to only run the query for the total count once for the first page requested and then cache it in some way for the rest of the pages, but that would be quite complicated and might also be difficult to invalidate properly (given the fact that hydrusrv doesn't know when hydrusrv-sync runs). So I'd leave that out, at least for now.~~
__Edit:__ see [here](https://github.com/mserajnik/hydrusrv/issues/15#issuecomment-444567465) for a promising solution to caching.

This feature will also need to be reflected and consumed in hydrusrvue, e.g., with a label like it's done on the _sorting_ field:

![result_count](https://user-images.githubusercontent.com/28833155/49524259-b7b52c00-f8ab-11e8-8a23-73ecd8abc5d2.png)

Also, when the total count is known, hydrusrvue will no longer need to make the final page request (which returns 0 results) to determine when it has fetched all the pages for a given request like it does currently. Therefore, we should also make use of this ability when we implement it in hydrusrvue.

---

In both cases, it could be set via environment variable, like it is now for most settings.

However, another option for this would be adding an additional query parameter (e.g., `getTotalCount=true`) to tell hydrusrv when it should calculate and respond the count. It could then be configurable on a per-user basis in hydrusrvue (via settings) ~~and hydrusrvue could do the caching of the count client-side (but, again, this can probably not be done in a _perfect_ way if files/tags are added/removed regularly, so I'm not sure if caching should be done at all)~~.

hydrusrv should still feature an environment variable flag to ignore the query parameter in case an administrator wants to disable it for all users. Likewise, an environment variable flag in hydrusrvue could be added to hide the setting from the user if getting the count is disabled in hydrusrv as well.

The first one would be easier to implement but the second one is probably the more user-friendly solution.

@FredericaBernkastel Any opinion on this?

**Comment by FredericaBernkastel** [comment]
Cool! And yes, it should be configurable, because it will likely slow down queries as twice.
I don't see any issues with caching, though. Just make a table `sha256(search tags) = count`. hydrusrv-sync will flush it on update.

**Comment by FredericaBernkastel** [comment]
I mean hash of the query as-is, without sorting and pagination parts
>`"tags[0]=loli&tags[1]=-rating:safe"`

**Comment by FredericaBernkastel** [comment]
What, no. It should work the same for everyone.

**Comment by FredericaBernkastel** [comment]
Admin will decide that, depends on the available hardware.

**Comment by mserajnik** [comment]
Alright. Sorry, wasn't sure.

> And yes, it should be configurable, because it will likely slow down queries as twice.

I read that as "configurable per user" because I took it as a given that it should be configurable "at all".

> Admin will decide that, depends on the available hardware.

He would be able to in the second variant as well though since he could disable it altogether (as described).

But generally, I agree, the second variant is complicated (both to develop and to understand from a users' view), so let's keep it simple (configuration only on the application level).

**Comment by mserajnik** [comment]
WIP: [`feature/counts`](https://github.com/mserajnik/hydrusrv/tree/feature/counts)

Basically done, but no caching yet. Performance is still pretty good, even on large DBs (like Roberts') _if_ the total set is not huge (10k+ files). E.g., a query with 150k results took around 250ms without count and 400ms with.

Next, I'll add support in hydrusrvue and we'll see how this goes from there.

**Comment by mserajnik** [comment]
I've added the caching as well now (same branch as before): https://github.com/mserajnik/hydrusrv/commit/5aab1aa001f72fe8daad15e60195cb4ac6fd5cd5

I've made it optional since from what I can see, on small databases there's basically no performance difference and the user might decide that it's not even worth the write operation it needs to do.

However, makes a big difference (as expected) for large databases, especially when running queries that have a lot of results.

Needs [`feature/counts`](https://github.com/mserajnik/hydrusrv-sync/tree/feature/counts) from hydrusrv-sync to handle the two new tables.

hydrusrvue compatibility WIP probably tomorrow.

**Comment by FredericaBernkastel** [comment]
>not even worth the write operation it needs to do
Although you can store caches in-memory. To invalidate it, you will have single table with single row, containing last-update timestamp from hydrusrv-sync.

**Comment by mserajnik** [comment]
Not with SQLite since everything in-memory is only for the connection (and each worker has its own).

Memcached or Redis should work, but adding another layer just for this? Not worth it, considering it makes the setup even more complicated than it already is. Plus, we are probably talking speed improvements of sub 1ms in this case.

**Comment by mserajnik** [comment]
hydrusrvue support: [`feature/counts`](https://github.com/mserajnik/hydrusrvue/tree/feature/counts)

As mentioned in the opening comment, hydrusrvue also utilizes the count (if available) to determine if it needs to fetch the next page or not by checking if the amount of tags/files it has stores matches the total count (which results in the final query which would return 0 results not being made). When this is not possible (when the user opens a different page than 1 directly), it falls back to the system used when no count is available.

Also, I've changed the lazy loading to fetch the next page slightly earlier than before to improve UX when scrolling. Until now, you'd have to hit the bottom of the page but with this change it fetches the next page once you hit exceed the `3rem` padding. Imo, this feels way better; when the query is fast, the user doesn't even notice it's loading additional pages (except from the scrollbar that gets bigger).

**Comment by mserajnik** [comment]
If you can, give it a try please and let me know if you find any issues. I did some testing and didn't find anything (aside from a small bug I've already fixed in the latest commit on the [hydrusrv feature/counts branch](https://github.com/mserajnik/hydrusrv/tree/feature/counts)).

Other than that, I think this works fine and is a nice addition.

**Comment by mserajnik** [comment]
> Also, i noticed, there is a support of "manga viewer" mode, but it's not quite finished. Is it possible to fetch next page once user hits end of the prevous one?

Unfortunately, not easily. I've already tried that when I first implemented it. The package I'm using doesn't support any hooks for that and even when I tried writing my own PhotoSwipe (the lightbox that's used) wrapper for Vue, I didn't find a way to implement it without causing all sorts of issues. For now, I'd like to keep this as is.

**Comment by mserajnik** [comment]
I wanted to say I'm sure there's a better fitting icon available in Font Awesome, but… I honestly can't find one. One that would kinda fit for the files search is [`th`](https://fontawesome.com/icons/th?style=solid) since it kinda looks like the grid, but this couldn't be used for tags and I don't want to have two different icons for the same meaning.

So yeah, I guess going with [`equals`](https://fontawesome.com/icons/equals?style=solid) makes the most sense. But it's weird to me that they don't have a single icon that can be used for indicating a count.

Btw, if we add an icon here, what do you think about the following as well (sorting and sorting direction):

<img width="1418" alt="screenshot 2018-12-08 at 20 42 51" src="https://user-images.githubusercontent.com/28833155/49690028-69619080-fb2a-11e8-9a7f-a9fb6859ee18.png">

<img width="417" alt="screenshot 2018-12-08 at 20 43 52" src="https://user-images.githubusercontent.com/28833155/49690030-6bc3ea80-fb2a-11e8-81b6-32b36ba24ebb.png">

Then every input label would have an icon instead of text (so it's consistent) and I think those are easy enough to understand.

**Comment by FredericaBernkastel** [comment]
Then you will have 2 problems:
1. icon should follow the direction (ascending/descending), otherwise it will confuse even more.
2. icon should follow type of the sort:
`"sort-numeric-down", "sort-numeric-up"` for id
`"sort-amount-up", "sort-amount-down"` for filesize, width and height
`"sort-alpha-up", "sort-alpha-down"` for namespaces
`"random"` for random

**Comment by mserajnik** [comment]
That's actually a neat idea. And easy to implement. I'll give it a try.

But I think just using `sort-alpha` for namespaces is enough to understand it (most non-technical people would think alphabetical sort equals natural sort anyway), two icons confuse even more and I wouldn't want a custom icon that doesn't come from Font Awesome.

**Comment by mserajnik** [comment]
Mime gets `sort-alpha` too, it's sorted alphabetically (and you can sort up/down).

**Comment by mserajnik** [comment]
Done, give it a try if you want: https://github.com/mserajnik/hydrusrvue/commit/e7ed0bfc8b6ecfd7071c99db76ee6ee9903217cc

**Comment by mserajnik** [comment]
Okay, I've just noticed something I don't like because it's a bit confusing:

Look at `sort-amount-up`. The arrow points up and the biggest bar is at the top.

Now look at `sort-alpha-up` or `sort-numeric-up`. The arrow points up but the `a` or `1` is at the top while the `z` or `9` is at the bottom.

Now, I know you kinda have to see the arrow and the icon besides it not as a whole unit but rather as "we are sorting in this direction using numeric sorting". But that might still confuse to people. And reversing the arrows for numeric and alphabetic sorting so the arrow points in the "correct direction" is also stupid.

Instead, we could just use [`long-arrow-alt-up`](https://fontawesome.com/icons/long-arrow-alt-up?style=solid) and [`long-arrow-alt-down`](https://fontawesome.com/icons/long-arrow-alt-down?style=solid) instead and [`arrows-alt-v`](https://fontawesome.com/icons/arrows-alt-v?style=solid) for the direction icon (where you choose it) since that fits perfectly to those two instead of `sort`. Much less confusing imo, the user only needs to know "up" or "down" anyway, no need for details like numeric or alphabetic sorting (he didn't has this info currently either). Plus, it saves some filesize since we don't need to load so many different icons.

What do you think?

**Comment by FredericaBernkastel** [comment]
I think you messed up with direction of sort-numeric.

**Comment by mserajnik** [comment]
Nope, I didn't. Read my last comment. :D

**Comment by FredericaBernkastel** [comment]
Poor fontawesome. I'd want to fix these icons manually, and attach it as a regular svg font.

**Comment by mserajnik** [comment]
Well, I guess their reasoning is the one I've mentioned previously (that the arrow indicates the direction and the "second icon" just the type of sorting without being "attached" to the arrow).

Glyphicons do it better (but I don't really like them visually):

<img width="805" alt="screenshot 2018-12-09 at 00 16 58" src="https://user-images.githubusercontent.com/28833155/49691660-cae42800-fb47-11e8-9202-39535176652a.png">

Regarding manually fixed icons: I appreciate it, but the goal for hydrusrvue is to use as little custom stuff as possible (aside from the few necessary CSS additions I had to make) to keep maintenance low. If we started using a few custom icons just for the sorting, we'd have to keep them up to date with any changes Font Awesome or Bulma (regarding the handling) might introduce in the future (so they don't start looking out of place). Too big of a hassle for what it's worth. Plus, I like having everything from a single source (in this case, Font Awesome for icons).

Regardless, I like the simpler solution (with just the arrows) better anyway (looks cleaner as well imo), so I'm going to switch to that, update the hydrusrvue screenshots (since there've been quite a few changes, your thumbnail adjustments as well) and prepare everything for new releases:

+ hydrusrv `5.0.0`
+ hydrusrv-sync `2.0.0`
+ hydrusrvue `2.0.0`

Major versions for all three to indicate incompatibility between them when using older versions. I'll explicitly mention this in their readmes as well.

Of course, hydrusrv-docker will get a new release as well, but just a minor version bump, since from the users POV nothing really changed for that aside from a few environment variables the user will have to adjust.

As always, thanks for your input! 👍
